### ISSUE TITLE
Accuracy calculation blowing up memory use

### ISSUE THREAD

**Issue by surya-narayanan** [Lightning-AI/torchmetrics#1738]
### To Reproduce

Steps to reproduce the behavior...

<details>
  <summary>Code sample</summary>

```python
import torch
from torchmetrics import Accuracy

print(torch.cuda.memory_allocated()/1024**2)
# 0.0

device = "cuda"


x = torch.randn(8, 512, 30522, device=device)
preds = x.argmax(dim = -1)
y = torch.randint(0, 30522, (8, 512), device=device)
print(torch.cuda.memory_allocated()/1024**2)
# 476.96875

for i in range(100):
  gen_accuracy = Accuracy(task="multiclass", num_classes=30522, ignore_index=-100, average = 'micro', multidim_average = 'samplewise').to(device)
  acc1 = torch.sum((preds == y).float(), dim = -1) #doesn't blow up memory
  acc2 = gen_accuracy(preds, y) #does blow up memory
  print(torch.cuda.memory_allocated()/1024**2)
```

</details>

### Expected behavior

Memory footprint stays constant and does not increase

### Environment

- TorchMetrics version (and how you installed TM, e.g. `conda`, `pip`, build from source):
- Python & PyTorch Version (e.g., 1.0):
- Any other relevant information such as OS (e.g., Linux):

Running this on google colab with a `!pip install torchmetrics` right before should be sufficient

**Comment by surya-narayanan** [comment]
additionally, this bug only seems to happen when multi-dim averaging is set to `samplewise`

**Comment by SkafteNicki** [comment]
@surya-narayanan thanks for reporting this issue. I try to figure out what is going on with the metric or if it is a general problem.
For now I can give you a stupid work around: simply add
```python
del gen_accuracy.tp, gen_accuracy.fp, gen_accuracy.tn, gen_accuracy.fn
```
to the end of your loop. That will fix the issue. 
This also indicates that it is some weird reference to metric states that are not being cleared as it should, and the memory is therefore not released at the end of the loop.

**Comment by SkafteNicki** [comment]
While tracking down what is going on here, I am also going to add that the comparison is not fair because for the code provided you are only calculating the value for the current batch and not utilizing the accumulation of the metric class at all. A better comparison would be with the functional version e.g.:
```python
for i in range(20):
    acc1 = torch.sum((preds == y).float(), dim = -1) #doesn't blow up memory
    acc2 = accuracy(preds, y, task="multiclass", num_classes=30522, ignore_index=-100, average = 'micro', multidim_average = 'samplewise') #doesn't blow up memory
    print(torch.cuda.memory_allocated()/1024**2)
```

**Comment by surya-narayanan** [comment]
Wow, thanks Nicki. I was wondeirng, what did you mean by "_not utilizing the accumulation of the metric class_"? Did you mean I wasn't doing something like 

```
import torch
from torchmetrics import Accuracy

print(torch.cuda.memory_allocated()/1024**2)
# 0.0

device = "cuda"


x = torch.randn(8, 512, 30522, device=device)
preds = x.argmax(dim = -1)
y = torch.randint(0, 30522, (8, 512), device=device)
print(torch.cuda.memory_allocated()/1024**2)
# 476.96875

myacc = 0

for i in range(100):
  gen_accuracy = Accuracy(task="multiclass", num_classes=30522, ignore_index=-100, average = 'micro', multidim_average = 'samplewise').to(device)
  acc1 = torch.sum((preds == y).float(), dim = -1) 
  acc2 = gen_accuracy(preds, y) 
  myacc += acc2 # is this what you meant by accumulation of the metric class?
  print(torch.cuda.memory_allocated()/1024**2)
```

Also, will assume that this is unresolved, but let me know if otherwise :)

**Comment by justusschock** [comment]
just for visibility, I'm cross-posting comments from our offline discussion (between @SkafteNicki @Borda and me):

> I think the issue is that lists are owned by Python GC which cleans as lazy as possible and never knows in advance if an object might be reused (unless you leave the current scope or tell it that it’s save to remove with del) whereas torch memory is handled by a separate PyTorch GC on the cpp level. In theory it might retain the empty python object that is essentially a pointer to invalid memory then, but that likely doesn’t use as much storage which is why it isn’t visible

---

> unfortunately the __del__ in python isn’t guaranteed to be called either, but maybe we could add a self.reset() there (or even better set the states to None/delete them) to at least help python with gc

---

> After looking at this for some time, I did manage narrow down what is causing this:
https://github.com/Lightning-AI/torchmetrics/blob/3c0d8a216d26a06a5fb3b85f187db8c6ed9f6e35/src/torchmetrics/metric.py#L562-L565
If we instead of storing and returning self._computed just returns the value the memory issue disappear. This would of cause come at the cost of the caching of the latest computed value. So two options as I see it:
solve the memory issue, but give up caching such that repeated calls to compute (with no change in-between) means that computations will be repeated
keep things as it is, and explain in the issue that the behavior is expected based on how garbage collection works in python
Both are corner cases, because the caching only has an effect if a user does something like:
metric.compute().compute().compute()
which nobody should need and the memory issue is only a problem if a user keeps initializing the same metric over and over again, which again should not be needed.

---

@surya-narayanan this is not yet resolved, but we know what's causing this, just not sure yet, what's the preferred way to fix since we don't control python's gc mechanism and don't want to force garbage collection if not necessary (this can cause substantial slowdowns). We'll keep this issue updated :)

**Comment by surya-narayanan** [comment]
Thanks
